File: ios/RecipeClipper/UI/Clip/ClipViewModel.swift
import Foundation
import Observation

/// What the snackbar says. The view picks the words; `.assigned` and `.cleared` offer Undo.
enum ClipMessage: Equatable {
    case assigned(ClipField, count: Int)
    case cleared(ClipField)
    /// A session draft for this page was brought back; offers Discard.
    case draftRestored
    case saveFailed
}

/// A `ClipMessage` to show once. `serial` tells two identical messages apart.
struct ClipNotice: Equatable {
    let message: ClipMessage
    let serial: Int
}

/// Android's ClipUiState. `selection` is the page's current selection split the way it would be
/// assigned. `newMarkId` is the mark the page should take from its current selection.
/// `savedRecipeId` is set once Save lands, so the view can open it.
struct ClipUiState: Equatable {
    let url: String
    var draft: ClipDraft
    var selection: [String] = []
    var pickingPhoto = false
    var newMarkId: String? = nil
    var reviewing = false
    var saving = false
    var notice: ClipNotice? = nil
    var savedRecipeId: Int64? = nil
}

/// "Clip it yourself" (#37), Android's ClipViewModel. The page itself lives in the view layer;
/// this only hears what happened on it (a selection, a tag or image tapped) and holds the
/// `ClipDraft`, kept in `ClipDraftStore` as it changes so Cancel keeps it for the session.
@MainActor
@Observable
final class ClipViewModel {
    private(set) var uiState: ClipUiState

    @ObservationIgnored private let url: String
    @ObservationIgnored private let repository: RecipeRepository
    @ObservationIgnored private let drafts: ClipDraftStore
    // The page's selection as given, before splitting: a name joins it rather than splitting it.
    @ObservationIgnored private var selectionText = ""
    // The draft before the last assignment or clear, for the snackbar's Undo.
    @ObservationIgnored private var undoTo: ClipDraft?
    @ObservationIgnored private var serial = 0

    init(url: String, repository: RecipeRepository, drafts: ClipDraftStore) {
        let cleaned = UrlCleaner.clean(url)
        self.url = cleaned
        self.repository = repository
        self.drafts = drafts
        let restored = drafts.get(cleaned)
        uiState = ClipUiState(url: cleaned, draft: restored ?? ClipDraft(sourceUrl: cleaned))
        if restored != nil { uiState.notice = notice(.draftRestored) }
    }

    // MARK: Events from the page

    func onSelectionChanged(_ text: String) {
        selectionText = text
        let lines = ClipSelection.lines(text)
        uiState.selection = lines
        // A new selection is never the one the last assignment was taken from.
        if !lines.isEmpty { uiState.newMarkId = nil }
    }

    /// Tapping a field's tag on the page clears that field, with Undo.
    func onTagTapped(_ field: ClipField) {
        let draft = uiState.draft
        guard draft.count(field) > 0 else { return }
        undoTo = draft
        setDraft(draft.clear(field), newMarkId: nil, message: .cleared(field))
    }

    /// While picking a photo, the next image tapped on the page becomes the photo.
    func onImageTapped(_ src: String) {
        guard uiState.pickingPhoto else { return }
        uiState.pickingPhoto = false
        assign(.photo, src)
    }

    // MARK: Events from the toolbar

    /// Puts the current selection into `field`, replacing what it held.
    func onAssign(_ field: ClipField) {
        guard field != .photo else { return }
        assign(field, selectionText)
    }

    func onPhotoButton() { uiState.pickingPhoto.toggle() }

    func onUndo() {
        guard let previous = undoTo else { return }
        undoTo = nil
        setDraft(previous, newMarkId: nil)
    }

    /// Throws the session draft away and starts over on the same page.
    func onDiscardDraft() {
        undoTo = nil
        drafts.remove(url)
        setDraft(ClipDraft(sourceUrl: url), newMarkId: nil)
    }

    func onNoticeShown(_ serial: Int) {
        if uiState.notice?.serial == serial { uiState.notice = nil }
    }

    // MARK: Review

    func onReview() {
        guard uiState.draft.canFinish else { return }
        uiState.reviewing = true
        uiState.pickingPhoto = false
    }

    func onBackToPage() { uiState.reviewing = false }

    func onNameChange(_ text: String) { edit { $0.name = text } }
    func onServesChange(_ text: String) { edit { $0.serves = text } }
    func onTotalTimeChange(_ text: String) { edit { $0.totalTime = text } }
    func onLineChange(_ field: ClipField, _ index: Int, _ text: String) { edit { $0 = $0.editLine(field, index, text) } }
    func onLineRemove(_ field: ClipField, _ index: Int) { edit { $0 = $0.removeLine(field, index) } }
    func onLineAdd(_ field: ClipField) { edit { $0 = $0.addLine(field) } }
    func onRemovePhoto() { edit { $0 = $0.clear(.photo) } }

    func onSave() {
        guard !uiState.saving, let recipe = uiState.draft.toRecipe() else { return }
        uiState.saving = true
        Task { [weak self, repository] in
            let result = await repository.saveClip(recipe)
            guard let self else { return }
            uiState.saving = false
            switch result {
            case .success(let saved):
                drafts.remove(url)
                uiState.savedRecipeId = saved.id
            case .error:
                uiState.notice = notice(.saveFailed)
            }
        }
    }

    // MARK: Internals

    private func assign(_ field: ClipField, _ text: String) {
        let draft = uiState.draft
        let markId = draft.pendingMarkId
        let assigned = draft.assign(field, text)
        guard assigned != draft else { return }
        undoTo = draft
        selectionText = ""
        uiState.selection = []
        setDraft(
            assigned,
            newMarkId: field == .photo ? nil : markId,
            message: .assigned(field, count: assigned.count(field))
        )
    }

    /// A hand edit in Review. Not undoable from the snackbar, so it ends any pending Undo.
    private func edit(_ change: (inout ClipDraft) -> Void) {
        undoTo = nil
        var draft = uiState.draft
        change(&draft)
        setDraft(draft, newMarkId: uiState.newMarkId)
    }

    private func setDraft(_ draft: ClipDraft, newMarkId: String?, message: ClipMessage? = nil) {
        drafts.put(draft)
        var state = uiState
        state.draft = draft
        state.newMarkId = newMarkId
        if let message { state.notice = notice(message) }
        // Discarding everything leaves nothing to review.
        state.reviewing = state.reviewing && !draft.isEmpty
        uiState = state
    }

    private func notice(_ message: ClipMessage) -> ClipNotice {
        serial += 1
        return ClipNotice(message: message, serial: serial)
    }
}

File: app/src/main/java/com/example/recipeclipper/ui/clip/ClipViewModel.kt
package com.example.recipeclipper.ui.clip

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.recipeclipper.data.ClipDraftStore
import com.example.recipeclipper.data.Entitlements
import com.example.recipeclipper.data.PurchaseOutcome
import com.example.recipeclipper.data.RecipeRepository
import com.example.recipeclipper.data.needsNotice
import com.example.recipeclipper.data.model.ClipDraft
import com.example.recipeclipper.data.model.ClipField
import com.example.recipeclipper.data.model.ClipSelection
import com.example.recipeclipper.data.model.ParseResult
import com.example.recipeclipper.data.model.UrlCleaner
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/** What the snackbar says. The screen picks the words; [Assigned] and [Cleared] offer Undo. */
sealed class ClipMessage {
    data class Assigned(val field: ClipField, val count: Int) : ClipMessage()
    data class Cleared(val field: ClipField) : ClipMessage()

    /** A session draft for this page was brought back; offers Discard. */
    object DraftRestored : ClipMessage()
    object SaveFailed : ClipMessage()

    /** The Unlock from the full-library prompt (#107) is pending or failed. */
    data class Unlock(val outcome: PurchaseOutcome) : ClipMessage()
}

/** A [ClipMessage] to show once. [serial] tells two identical messages apart. */
data class ClipNotice(val message: ClipMessage, val serial: Long)

/**
 * [selection] is the page's current selection split the way it would be assigned (one item a
 * line). [newMarkId] is the mark the page should take from its current selection: the id the
 * last assignment recorded. [savedRecipeId] is set once Save lands, so the screen can open it.
 */
data class ClipUiState(
    val url: String,
    val draft: ClipDraft,
    val selection: List<String> = emptyList(),
    val pickingPhoto: Boolean = false,
    val newMarkId: String? = null,
    val reviewing: Boolean = false,
    val saving: Boolean = false,
    val notice: ClipNotice? = null,
    val savedRecipeId: Long? = null,
    /** The clip couldn't be saved: the free library is full and all protected (#107). */
    val libraryFull: Boolean = false
)

/**
 * "Clip it yourself" (#37): the user selects text on a page with no recipe data and says where
 * each part goes. The page itself lives in the view layer; this only hears what happened on it
 * (a selection, a tag or image tapped) and holds the [ClipDraft].
 *
 * The draft is kept in [ClipDraftStore] as it changes, so Cancel keeps it for the session, and
 * mirrored into [SavedStateHandle] so it survives the process being killed in the background.
 */
@HiltViewModel
class ClipViewModel @Inject constructor(
    private val savedStateHandle: SavedStateHandle,
    private val repository: RecipeRepository,
    private val drafts: ClipDraftStore,
    private val entitlements: Entitlements = Entitlements.Unavailable
) : ViewModel() {

    private val url: String = UrlCleaner.clean(savedStateHandle.get<String>(URL_ARG).orEmpty())

    private val _uiState: MutableStateFlow<ClipUiState>
    val uiState: StateFlow<ClipUiState>

    // The page's selection as given, before splitting: a name joins it rather than splitting it.
    private var selectionText = ""

    // The draft before the last assignment or clear, for the snackbar's Undo.
    private var undoTo: ClipDraft? = null

    private var serial = 0L

    init {
        val restored = drafts.get(url) ?: restoreFrom(savedStateHandle)
        _uiState = MutableStateFlow(
            ClipUiState(
                url = url,
                draft = restored ?: ClipDraft(url),
                notice = restored?.let { notice(ClipMessage.DraftRestored) }
            )
        )
        uiState = _uiState.asStateFlow()
    }

    // --- Events from the page ---

    fun onSelectionChanged(text: String) {
        selectionText = text
        val lines = ClipSelection.lines(text)
        _uiState.update {
            // A new selection is never the one the last assignment was taken from.
            it.copy(selection = lines, newMarkId = if (lines.isEmpty()) it.newMarkId else null)
        }
    }

    /** Tapping a field's tag on the page clears that field, with Undo. */
    fun onTagTapped(field: ClipField) {
        val draft = _uiState.value.draft
        if (draft.count(field) == 0) return
        undoTo = draft
        setDraft(draft.clear(field), newMarkId = null, message = ClipMessage.Cleared(field))
    }

    /** While picking a photo, the next image tapped on the page becomes the photo. */
    fun onImageTapped(src: String) {
        if (!_uiState.value.pickingPhoto) return
        _uiState.update { it.copy(pickingPhoto = false) }
        assign(ClipField.PHOTO, src)
    }

    // --- Events from the toolbar ---

    /** Puts the current selection into [field], replacing what it held. */
    fun onAssign(field: ClipField) {
        if (field == ClipField.PHOTO) return
        assign(field, selectionText)
    }

    fun onPhotoButton() {
        _uiState.update { it.copy(pickingPhoto = !it.pickingPhoto) }
    }

    fun onUndo() {
        val previous = undoTo ?: return
        undoTo = null
        setDraft(previous, newMarkId = null)
    }

    /** Throws the session draft away and starts over on the same page. */
    fun onDiscardDraft() {
        undoTo = null
        drafts.remove(url)
        setDraft(ClipDraft(url), newMarkId = null)
    }

    fun onNoticeShown(serial: Long) {
        _uiState.update { if (it.notice?.serial == serial) it.copy(notice = null) else it }
    }

    // --- Review ---

    fun onReview() {
        if (_uiState.value.draft.canFinish) _uiState.update { it.copy(reviewing = true, pickingPhoto = false) }
    }

    fun onBackToPage() = _uiState.update { it.copy(reviewing = false) }

    fun onNameChange(text: String) = edit { it.copy(name = text) }
    fun onServesChange(text: String) = edit { it.copy(serves = text) }
    fun onTotalTimeChange(text: String) = edit { it.copy(totalTime = text) }
    fun onLineChange(field: ClipField, index: Int, text: String) = edit { it.editLine(field, index, text) }
    fun onLineRemove(field: ClipField, index: Int) = edit { it.removeLine(field, index) }
    fun onLineAdd(field: ClipField) = edit { it.addLine(field) }
    fun onRemovePhoto() = edit { it.clear(ClipField.PHOTO) }

    fun onSave() {
        val state = _uiState.value
        if (state.saving) return
        val recipe = state.draft.toRecipe() ?: return
        _uiState.update { it.copy(saving = true) }
        viewModelScope.launch {
            when (val result = repository.saveClip(recipe)) {
                is ParseResult.Success -> if (!result.kept) {
                    _uiState.update { it.copy(saving = false, libraryFull = true) }
                } else {
                    drafts.remove(url)
                    clearSavedState()
                    _uiState.update { it.copy(saving = false, savedRecipeId = result.recipe.id) }
                }
                is ParseResult.Error -> _uiState.update {
                    it.copy(saving = false, notice = notice(ClipMessage.SaveFailed))
                }
            }
        }
    }

    /** Unlock from the full-library prompt (#107), then save the clip as it stands. */
    fun onUnlock() {
        _uiState.update { it.copy(libraryFull = false) }
        viewModelScope.launch {
            val outcome = entitlements.purchase()
            if (outcome == PurchaseOutcome.UNLOCKED) onSave()
            else if (outcome.needsNotice) _uiState.update { it.copy(notice = notice(ClipMessage.Unlock(outcome))) }
        }
    }

    fun onLibraryFullDismiss() = _uiState.update { it.copy(libraryFull = false) }

    // --- Internals ---

    private fun assign(field: ClipField, text: String) {
        val draft = _uiState.value.draft
        val markId = draft.pendingMarkId
        val assigned = draft.assign(field, text)
        if (assigned === draft) return
        undoTo = draft
        selectionText = ""
        _uiState.update { it.copy(selection = emptyList()) }
        setDraft(
            assigned,
            newMarkId = if (field == ClipField.PHOTO) null else markId,
            message = ClipMessage.Assigned(field, assigned.count(field))
        )
    }

    /** A hand edit in Review. Not undoable from the snackbar, so it ends any pending Undo. */
    private fun edit(change: (ClipDraft) -> ClipDraft) {
        undoTo = null
        setDraft(change(_uiState.value.draft), newMarkId = _uiState.value.newMarkId)
    }

    private fun setDraft(draft: ClipDraft, newMarkId: String?, message: ClipMessage? = null) {
        drafts.put(draft)
        saveState(draft)
        _uiState.update {
            it.copy(
                draft = draft,
                newMarkId = newMarkId,
                notice = message?.let(::notice) ?: it.notice,
                // Discarding everything leaves nothing to review.
                reviewing = it.reviewing && !draft.isEmpty
            )
        }
    }

    private fun notice(message: ClipMessage) = ClipNotice(message, ++serial)

    private fun saveState(draft: ClipDraft) {
        if (draft.isEmpty) return clearSavedState()
        savedStateHandle[KEY_NAME] = draft.name
        savedStateHandle[KEY_INGREDIENTS] = ArrayList(draft.ingredients)
        savedStateHandle[KEY_STEPS] = ArrayList(draft.steps)
        savedStateHandle[KEY_PHOTO] = draft.photo
        savedStateHandle[KEY_SERVES] = draft.serves
        savedStateHandle[KEY_TOTAL_TIME] = draft.totalTime
    }

    private fun clearSavedState() {
        ALL_KEYS.forEach { savedStateHandle.remove<Any>(it) }
    }

    /** The draft mirrored before the process died. Its page marks are gone with the page. */
    private fun restoreFrom(handle: SavedStateHandle): ClipDraft? {
        if (!handle.contains(KEY_NAME)) return null
        return ClipDraft(
            sourceUrl = url,
            name = handle.get<String>(KEY_NAME).orEmpty(),
            ingredients = handle.get<ArrayList<String>>(KEY_INGREDIENTS).orEmpty(),
            steps = handle.get<ArrayList<String>>(KEY_STEPS).orEmpty(),
            photo = handle.get<String>(KEY_PHOTO),
            serves = handle.get<String>(KEY_SERVES).orEmpty(),
            totalTime = handle.get<String>(KEY_TOTAL_TIME).orEmpty()
        ).takeUnless { it.isEmpty }
    }

    companion object {
        const val URL_ARG = "url"
        private const val KEY_NAME = "clip.name"
        private const val KEY_INGREDIENTS = "clip.ingredients"
        private const val KEY_STEPS = "clip.steps"
        private const val KEY_PHOTO = "clip.photo"
        private const val KEY_SERVES = "clip.serves"
        private const val KEY_TOTAL_TIME = "clip.totalTime"
        private val ALL_KEYS = listOf(KEY_NAME, KEY_INGREDIENTS, KEY_STEPS, KEY_PHOTO, KEY_SERVES, KEY_TOTAL_TIME)
    }
}
